Text not entered in to Stripe iFrame with Safari

<!--
If you have all reproduction steps with a complete sample app, please share as many details as possible in the sections below.

Make sure that you tried using the latest TestCafe version (https://github.com/DevExpress/testcafe/releases), where this behavior might have been already addressed.

Before submitting an issue, please check CONTRIBUTING.md and existing issues in this repository (https://github.com/DevExpress/testcafe/issues) in case a similar issue exists or was already addressed. This may save your time (and ours).
-->

### What is your Test Scenario?
<!-- Describe what you'd like to test. -->
 I'm trying to run our test code on multiple browsers, specifically entering in CreditCard details on our payment pages.

### What is the Current behavior?
<!-- Describe the behavior you see and consider invalid. -->
It looks like text isn't being properly entered in to the Stripe payment iframes, on Safari. The same steps are successful in Chrome, Firefox and Edge.

The main concern here is that TestCafe continues on, believing it has actually entered text in to these iframes.
 
### What is the Expected behavior?
<!-- Describe what you expected to happen. -->
 Text should be entered in to Stripe iFrames, regardless of browser. OR, at the least, if TestCafe is unable to locate/interact with the iFrame... it should fail at that point instead of continuing on.

### What is your web application and your TestCafe test code?
<!-- Share a public accessible link to your application or provide a simple app which we can run. -->

<details>
<summary>Your website URL (or attach your complete example): </summary>

<!-- Provide your website URL or attach a sample. Note: if your website requires any additional access procedures like authentication, please ask the website owner to send us a written confirmation at [support@devexpress.com](mailto:support@devexpress.com) in a free text form. It will allow the DevExpress staff to remotely access the website and its internal resources for research, testing, and debugging purposes. -->
You can target our live site, since the issue relates to being unable to enter data in the payment forms:
https://www.change.org/
</details>

<details>
<summary>Your complete test code (or attach your test files):</summary>

<!-- Paste your test code here: --> 
```js
import { t, Selector } from 'testcafe';

fixture('Join Membership with credit card').page('https://www.change.org/');

test('Guest user clicks contribute directly', async () => {
  await t
    .navigateTo('s/member')
    .expect(Selector('[data-testid="member_landing_page_inline_contribute_button"]').visible)
    .ok();

  await t
    .click(Selector('[data-testid="member_landing_page_inline_contribute_button"]'), { speed: 0.3 })
    .expect(Selector('[data-testid="member_payment_form"]').visible)
    .ok();

  await t
    .click(Selector('[data-testid="payment-option-button-creditCard'))
    .expect(Selector('.iframe-form-element').visible)
    .ok();

  const emailAddressInput = Selector('[data-testid="input_email"]').filterVisible();
  const confirmationEmailInput = Selector('[data-testid="input_confirmation_email"]').filterVisible();
  const firstNameInput = Selector('[data-testid="input_first_name"]');
  const lastNameInput = Selector('[data-testid="input_last_name"]');

  await t
    .typeText(emailAddressInput, 'email@email.com')
    .typeText(confirmationEmailInput, 'email@email.com')
    .typeText(firstNameInput, 'Your')
    .typeText(lastNameInput, 'Name');

  await t
    .switchToIframe(Selector('[data-testid="credit-card-number"] iframe'))
    .typeText(Selector('input[name="cardnumber"]'), '1234123412341234', { replace: true })
    .expect(Selector('input[name="cardnumber"]').value)
    .eql('1234 1234 1234 1234')
    .switchToMainWindow();
});
```
</details>

<details>
<summary>Your complete configuration file (if any):</summary>

<!-- Paste your complete test config file here (even if it is huge): -->
```
N/A
```
</details>

<details>
<summary>Your complete test report:</summary>

<!-- Paste your complete result test report here (even if it is huge): -->
```
Guest user clicks contribute directly

   1) AssertionError: expected '' to deeply equal '1234 1234 1234 1234'

      Browser: Safari 13.0.5 / macOS 10.15.3

         31 |
         32 |  await t
         33 |    .switchToIframe(Selector('[data-testid="credit-card-number"] iframe'))
         34 |    .typeText(Selector('input[name="cardnumber"]'), '1234123412341234', { replace: true
      })
         35 |    .expect(Selector('input[name="cardnumber"]').value)
       > 36 |    .eql('1234 1234 1234 1234')
         37 |    .switchToMainWindow();
         38 |});
         39 |

         at <anonymous>
      (/Users/rcooper/work/github.com/change/regression-qaa/tests/users/demo.js:36:6)



 1/1 failed (23s)
```
</details>

<details>
<summary>Screenshots:</summary>
<!-- If applicable, add screenshots to help explain the issue. -->

```
N/A
```
</details>

### Steps to Reproduce:
<!-- Describe what we should do to reproduce the behavior you encountered. -->

1. Go to my website ...
3. Execute this command...
4. See the error...
 
### Your Environment details:
 
* testcafe version:  1.8.2                 <!-- run `testcafe -v` -->
* node.js version: 12.14.1                   <!-- run `node -v` -->
* command-line arguments: testcafe safari <!-- example: "testcafe ie,chrome -e test.js" -->
* browser name and version: Safari 13.0.5 <!-- example: IE 11, Chrome 69, Firefox 100, etc. -->
* platform and version: macOS 10.15.3         <!-- example: "macOS 10.14, Windows, Linux Ubuntu 18.04.1, iOS 12 -->

I've tried running your test case in several browsers (Chrome, Firefox, Safari) and it fails at line 14 with the same reason. Are there any additional actions or configuration steps I need to perform in order to run your test case, so it does not fail?

```
 Running tests in:
 - Chrome 80.0.3987.116 / macOS 10.15.2

 Join Membership with credit card
 ✖ Guest user clicks contribute directly

   1) AssertionError: expected false to be truthy

      Browser: Chrome 80.0.3987.116 / macOS 10.15.2

          9 |    .ok();
         10 |
         11 |  await t
         12 |    .click(Selector('[data-testid="member_landing_page_inline_contribute_button"]'), { speed: 0.3 })
         13 |    .expect(Selector('[data-testid="member_payment_form"]').visible)
       > 14 |    .ok();
         15 |
         16 |  await t
         17 |    .click(Selector('[data-testid="payment-option-button-creditCard'))
         18 |    .expect(Selector('.iframe-form-element').visible)
         19 |    .ok();

         at <anonymous> (/Users/testuser/Desktop/skorkin/4793.js:14:6)



 1/1 failed (19s)

--

  Running tests in:
 - Safari 13.0.4 / macOS 10.15.2

 Join Membership with credit card
 ✖ Guest user clicks contribute directly

   1) AssertionError: expected false to be truthy

      Browser: Safari 13.0.4 / macOS 10.15.2

          9 |    .ok();
         10 |
         11 |  await t
         12 |    .click(Selector('[data-testid="member_landing_page_inline_contribute_button"]'), { speed: 0.3 })
         13 |    .expect(Selector('[data-testid="member_payment_form"]').visible)
       > 14 |    .ok();
         15 |
         16 |  await t
         17 |    .click(Selector('[data-testid="payment-option-button-creditCard'))
         18 |    .expect(Selector('.iframe-form-element').visible)
         19 |    .ok();

         at <anonymous> (/Users/testuser/Desktop/skorkin/4793.js:14:6)



 1/1 failed (16s)

 --

 Running tests in:
 - Firefox 73.0 / macOS 10.15

 Join Membership with credit card
 ✖ Guest user clicks contribute directly

   1) AssertionError: expected false to be truthy

      Browser: Firefox 73.0 / macOS 10.15

          9 |    .ok();
         10 |
         11 |  await t
         12 |    .click(Selector('[data-testid="member_landing_page_inline_contribute_button"]'), { speed: 0.3 })
         13 |    .expect(Selector('[data-testid="member_payment_form"]').visible)
       > 14 |    .ok();
         15 |
         16 |  await t
         17 |    .click(Selector('[data-testid="payment-option-button-creditCard'))
         18 |    .expect(Selector('.iframe-form-element').visible)
         19 |    .ok();

         at <anonymous> (/Users/testuser/Desktop/skorkin/4793.js:14:6)



 1/1 failed (21s)
```

@rob4629 

Hello,

I was able to reproduce it. There is no `t.typeText(Selector('input[name="cardnumber"]'), '1234123412341234', { replace: true })` action result in the `input` element (Safari browser). We will examine it and check for a suitable solution.

**For the team.**
Test:
```js
import { t, Selector } from 'testcafe';

fixture('Join Membership with credit card').page('https://www.change.org/');

test('Guest user clicks contribute directly', async () => {
    await t
        .navigateTo('s/member?country_code=US&lang=en-US')
        .expect(Selector('[data-testid="member_landing_page_inline_contribute_button"]').visible)
        .ok();

    await t
        .click(Selector('[data-testid="member_landing_page_inline_contribute_button"]'), { speed: 0.3 })
        .expect(Selector('[data-testid="member_payment_form"]').visible)
        .ok();

    await t
        .click(Selector('[data-testid="payment-option-button-creditCard'))
        .expect(Selector('.iframe-form-element').visible)
        .ok();

    const emailAddressInput = Selector('[data-testid="input_email"]').filterVisible();
    const confirmationEmailInput = Selector('[data-testid="input_confirmation_email"]').filterVisible();
    const firstNameInput = Selector('[data-testid="input_first_name"]');
    const lastNameInput = Selector('[data-testid="input_last_name"]');

    await t
        .typeText(emailAddressInput, 'email@email.com')
        .typeText(confirmationEmailInput, 'email@email.com')
        .typeText(firstNameInput, 'Your')
        .typeText(lastNameInput, 'Name');

    await t
        .switchToIframe(Selector('[data-testid="credit-card-number"] iframe'))
        .typeText(Selector('input[name="cardnumber"]'), '1234123412341234', { replace: true })
        .expect(Selector('input[name="cardnumber"]').value)
        .eql('1234 1234 1234 1234')
        .switchToMainWindow();
});

```

Command:
```
testcafe remote test.js --hostname XXX.XX.XX.XXX --ports 1337,1338 -e
```
Result:
```
Connecting 1 remote browser(s)...
Navigate to the following URL from each remote browser.
Connect URL: http://172.22.14.134:1337/browser/connect
CONNECTED Safari 13.0.4 / macOS 10.15.2
 Running tests in:
 - Safari 13.0.4 / macOS 10.15.2

 Join Membership with credit card
 × Guest user clicks contribute directly

   1) AssertionError: expected '' to deeply equal '1234 1234 1234 1234'

      Browser: Safari 13.0.4 / macOS 10.15.2

         31 |
         32 |    await t
         33 |        .switchToIframe(Selector('[data-testid="credit-card-number"] iframe'))
         34 |        .typeText(Selector('input[name="cardnumber"]'), '1234123412341234', { replace: true })
         35 |        .expect(Selector('input[name="cardnumber"]').value)
       > 36 |        .eql('1234 1234 1234 1234')
         37 |        .switchToMainWindow();
         38 |});
         39 |

         at <anonymous> (d:\tst\i4793\test.js:36:10)



 1/1 failed (20s)
```

--- a/src/client/automation/cursor.js
+++ b/src/client/automation/cursor.js
@@ -1,8 +1,9 @@
 import testCafeCore from './deps/testcafe-core';
 import testCafeUI from './deps/testcafe-ui';
+import isWindowInIframe from '../../utils/is-window-in-iframe';
 
 const domUtils = testCafeCore.domUtils;
-const cursorUI = window.top === window ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
+const cursorUI = !isWindowInIframe(window)() ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
 
 
 // NOTE: the default position should be outside of the page (GH-794)
@@ -40,7 +41,7 @@ export default {
     },
 
     get visible () {
-        return window.top === window && cursorUI.isVisible();
+        return !isWindowInIframe(window)() && cursorUI.isVisible();
     },
 
     move (newX, newY) {
@@ -56,7 +57,7 @@ export default {
     },
 
     show () {
-        if (window.top === window)
+        if (!isWindowInIframe(window)())
             cursorUI.show();
     },
 

--- a/src/client/automation/cursor.js
+++ b/src/client/automation/cursor.js
@@ -3,7 +3,7 @@ import testCafeUI from './deps/testcafe-ui';
 import isWindowInIframe from '../../utils/is-window-in-iframe';
 
 const domUtils = testCafeCore.domUtils;
-const cursorUI = !isWindowInIframe(window)() ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
+const cursorUI = !isWindowInIframe(window) ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
 
 
 // NOTE: the default position should be outside of the page (GH-794)
@@ -41,7 +41,7 @@ export default {
     },
 
     get visible () {
-        return !isWindowInIframe(window)() && cursorUI.isVisible();
+        return !isWindowInIframe(window) && cursorUI.isVisible();
     },
 
     move (newX, newY) {
@@ -57,7 +57,7 @@ export default {
     },
 
     show () {
-        if (!isWindowInIframe(window)())
+        if (!isWindowInIframe(window))
             cursorUI.show();
     },
 

--- a/src/client/automation/cursor.js
+++ b/src/client/automation/cursor.js
@@ -1,9 +1,9 @@
 import testCafeCore from './deps/testcafe-core';
 import testCafeUI from './deps/testcafe-ui';
-import isWindowInIframe from '../../utils/is-window-in-iframe';
+import isIframeWindow from '../../utils/is-window-in-iframe';
 
 const domUtils = testCafeCore.domUtils;
-const cursorUI = !isWindowInIframe(window) ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
+const cursorUI = !isIframeWindow(window) ? testCafeUI.cursorUI : testCafeUI.iframeCursorUI;
 
 
 // NOTE: the default position should be outside of the page (GH-794)
@@ -41,7 +41,7 @@ export default {
     },
 
     get visible () {
-        return !isWindowInIframe(window) && cursorUI.isVisible();
+        return !isIframeWindow(window) && cursorUI.isVisible();
     },
 
     move (newX, newY) {
@@ -57,7 +57,7 @@ export default {
     },
 
     show () {
-        if (!isWindowInIframe(window))
+        if (!isIframeWindow(window))
             cursorUI.show();
     },
 

--- a/src/client/automation/get-element.js
+++ b/src/client/automation/get-element.js
@@ -94,7 +94,7 @@ export function fromPoint (x, y, expectedElement) {
             // In this case, you should hide a top window's shadow-ui root to obtain an element.
             let resChain = Promise.resolve(topElement);
 
-            if (!foundElement && isWindowInIframe(window)() && x > 0 && y > 0) {
+            if (!foundElement && isWindowInIframe(window) && x > 0 && y > 0) {
                 resChain = resChain
                     .then(() => getElementFromPoint(x, y, true))
                     .then(element => {

--- a/src/client/automation/get-element.js
+++ b/src/client/automation/get-element.js
@@ -2,7 +2,7 @@ import hammerhead from './deps/hammerhead';
 import testCafeCore from './deps/testcafe-core';
 import testCafeUI from './deps/testcafe-ui';
 import cursor from './cursor';
-import isWindowInIframe from '../../utils/is-window-in-iframe';
+import isIframeWindow from '../../utils/is-window-in-iframe';
 
 const browserUtils  = hammerhead.utils.browser;
 const Promise       = hammerhead.Promise;
@@ -94,7 +94,7 @@ export function fromPoint (x, y, expectedElement) {
             // In this case, you should hide a top window's shadow-ui root to obtain an element.
             let resChain = Promise.resolve(topElement);
 
-            if (!foundElement && isWindowInIframe(window) && x > 0 && y > 0) {
+            if (!foundElement && isIframeWindow(window) && x > 0 && y > 0) {
                 resChain = resChain
                     .then(() => getElementFromPoint(x, y, true))
                     .then(element => {

--- a/src/client/automation/get-element.js
+++ b/src/client/automation/get-element.js
@@ -2,6 +2,7 @@ import hammerhead from './deps/hammerhead';
 import testCafeCore from './deps/testcafe-core';
 import testCafeUI from './deps/testcafe-ui';
 import cursor from './cursor';
+import isWindowInIframe from '../../utils/is-window-in-iframe';
 
 const browserUtils  = hammerhead.utils.browser;
 const Promise       = hammerhead.Promise;
@@ -82,7 +83,6 @@ function correctTopElementByExpectedElement (topElement, expectedElement) {
 }
 
 export function fromPoint (x, y, expectedElement) {
-    const isInIframe   = window !== window.top;
     let foundElement = null;
 
     return getElementFromPoint(x, y)
@@ -94,7 +94,7 @@ export function fromPoint (x, y, expectedElement) {
             // In this case, you should hide a top window's shadow-ui root to obtain an element.
             let resChain = Promise.resolve(topElement);
 
-            if (!foundElement && isInIframe && x > 0 && y > 0) {
+            if (!foundElement && isWindowInIframe(window)() && x > 0 && y > 0) {
                 resChain = resChain
                     .then(() => getElementFromPoint(x, y, true))
                     .then(element => {

--- a/src/client/automation/playback/move/index.js
+++ b/src/client/automation/playback/move/index.js
@@ -13,7 +13,7 @@ import AutomationSettings from '../../settings';
 import DragAndDropState from '../drag/drag-and-drop-state';
 import createEventSequence from './event-sequence/create-event-sequence';
 import lastHoveredElementHolder from '../last-hovered-element-holder';
-import isWindowInIframe from '../../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../../utils/is-window-in-iframe';
 
 const Promise          = hammerhead.Promise;
 const nativeMethods    = hammerhead.nativeMethods;
@@ -413,7 +413,7 @@ export default class MoveAutomation {
             .then(message => {
                 cursor.activeWindow = window;
 
-                if (iframeUnderCursor || isWindowInIframe(window))
+                if (iframeUnderCursor || isIframeWindow(window))
                     return cursor.move(message.x, message.y);
 
                 return null;

--- a/src/client/automation/playback/move/index.js
+++ b/src/client/automation/playback/move/index.js
@@ -13,6 +13,7 @@ import AutomationSettings from '../../settings';
 import DragAndDropState from '../drag/drag-and-drop-state';
 import createEventSequence from './event-sequence/create-event-sequence';
 import lastHoveredElementHolder from '../last-hovered-element-holder';
+import isWindowInIframe from '../../../../utils/is-window-in-iframe';
 
 const Promise          = hammerhead.Promise;
 const nativeMethods    = hammerhead.nativeMethods;
@@ -412,7 +413,7 @@ export default class MoveAutomation {
             .then(message => {
                 cursor.activeWindow = window;
 
-                if (iframeUnderCursor || window.top !== window)
+                if (iframeUnderCursor || isWindowInIframe(window)())
                     return cursor.move(message.x, message.y);
 
                 return null;

--- a/src/client/automation/playback/move/index.js
+++ b/src/client/automation/playback/move/index.js
@@ -413,7 +413,7 @@ export default class MoveAutomation {
             .then(message => {
                 cursor.activeWindow = window;
 
-                if (iframeUnderCursor || isWindowInIframe(window)())
+                if (iframeUnderCursor || isWindowInIframe(window))
                     return cursor.move(message.x, message.y);
 
                 return null;

--- a/src/client/automation/playback/press/index.js
+++ b/src/client/automation/playback/press/index.js
@@ -168,7 +168,7 @@ export default class PressAutomation {
         const activeElement         = domUtils.getActiveElement();
         const activeElementIsIframe = domUtils.isIframeElement(activeElement);
 
-        if (!isWindowInIframe(window)() && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
+        if (!isWindowInIframe(window) && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
             const msg = {
                 cmd:             PRESS_REQUEST_CMD,
                 keyCombinations: this.keyCombinations,

--- a/src/client/automation/playback/press/index.js
+++ b/src/client/automation/playback/press/index.js
@@ -11,6 +11,7 @@ import KeyPressSimulator from './key-press-simulator';
 import supportedShortcutHandlers from './shortcuts';
 import { getActualKeysAndEventKeyProperties, getDeepActiveElement } from './utils';
 import AutomationSettings from '../../settings';
+import isWindowInIframe from '../../../../utils/is-window-in-iframe';
 
 const Promise        = hammerhead.Promise;
 const browserUtils   = hammerhead.utils.browser;
@@ -167,7 +168,7 @@ export default class PressAutomation {
         const activeElement         = domUtils.getActiveElement();
         const activeElementIsIframe = domUtils.isIframeElement(activeElement);
 
-        if (window.top === window && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
+        if (!isWindowInIframe(window)() && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
             const msg = {
                 cmd:             PRESS_REQUEST_CMD,
                 keyCombinations: this.keyCombinations,

--- a/src/client/automation/playback/press/index.js
+++ b/src/client/automation/playback/press/index.js
@@ -11,7 +11,7 @@ import KeyPressSimulator from './key-press-simulator';
 import supportedShortcutHandlers from './shortcuts';
 import { getActualKeysAndEventKeyProperties, getDeepActiveElement } from './utils';
 import AutomationSettings from '../../settings';
-import isWindowInIframe from '../../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../../utils/is-window-in-iframe';
 
 const Promise        = hammerhead.Promise;
 const browserUtils   = hammerhead.utils.browser;
@@ -168,7 +168,7 @@ export default class PressAutomation {
         const activeElement         = domUtils.getActiveElement();
         const activeElementIsIframe = domUtils.isIframeElement(activeElement);
 
-        if (!isWindowInIframe(window) && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
+        if (!isIframeWindow(window) && activeElementIsIframe && nativeMethods.contentWindowGetter.call(activeElement)) {
             const msg = {
                 cmd:             PRESS_REQUEST_CMD,
                 keyCombinations: this.keyCombinations,

--- a/src/client/automation/playback/scroll.js
+++ b/src/client/automation/playback/scroll.js
@@ -260,7 +260,7 @@ export default class ScrollAutomation {
 
         return scrollParentsPromise
             .then(() => {
-                if (isWindowInIframe(window)() && !this.skipParentFrames) {
+                if (isWindowInIframe(window) && !this.skipParentFrames) {
                     return sendRequestToFrame({
                         cmd:             SCROLL_REQUEST_CMD,
                         offsetX:         currentOffsetX,

--- a/src/client/automation/playback/scroll.js
+++ b/src/client/automation/playback/scroll.js
@@ -7,7 +7,7 @@ import {
     scrollController,
     sendRequestToFrame
 } from '../deps/testcafe-core';
-import isWindowInIframe from '../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../utils/is-window-in-iframe';
 
 const Promise        = hammerhead.Promise;
 const messageSandbox = hammerhead.eventSandbox.message;
@@ -260,7 +260,7 @@ export default class ScrollAutomation {
 
         return scrollParentsPromise
             .then(() => {
-                if (isWindowInIframe(window) && !this.skipParentFrames) {
+                if (isIframeWindow(window) && !this.skipParentFrames) {
                     return sendRequestToFrame({
                         cmd:             SCROLL_REQUEST_CMD,
                         offsetX:         currentOffsetX,

--- a/src/client/automation/playback/scroll.js
+++ b/src/client/automation/playback/scroll.js
@@ -7,6 +7,7 @@ import {
     scrollController,
     sendRequestToFrame
 } from '../deps/testcafe-core';
+import isWindowInIframe from '../../../utils/is-window-in-iframe';
 
 const Promise        = hammerhead.Promise;
 const messageSandbox = hammerhead.eventSandbox.message;
@@ -259,7 +260,7 @@ export default class ScrollAutomation {
 
         return scrollParentsPromise
             .then(() => {
-                if (window.top !== window && !this.skipParentFrames) {
+                if (isWindowInIframe(window)() && !this.skipParentFrames) {
                     return sendRequestToFrame({
                         cmd:             SCROLL_REQUEST_CMD,
                         offsetX:         currentOffsetX,

--- a/src/client/automation/playback/scroll.js
+++ b/src/client/automation/playback/scroll.js
@@ -17,7 +17,6 @@ const SCROLL_MARGIN_INCREASE_STEP = 20;
 const SCROLL_REQUEST_CMD  = 'automation|scroll|request';
 const SCROLL_RESPONSE_CMD = 'automation|scroll|response';
 
-
 // Setup cross-iframe interaction
 messageSandbox.on(messageSandbox.SERVICE_MSG_RECEIVED_EVENT, e => {
     if (e.message.cmd === SCROLL_REQUEST_CMD) {

--- a/src/client/automation/utils/utils.js
+++ b/src/client/automation/utils/utils.js
@@ -1,7 +1,7 @@
 import hammerhead from '../deps/hammerhead';
 import testCafeCore from '../deps/testcafe-core';
 import sendRequestToFrame from '../../core/utils/send-request-to-frame';
-import isWindowInIframe from '../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../utils/is-window-in-iframe';
 
 const Promise          = hammerhead.Promise;
 const nativeMethods    = hammerhead.nativeMethods;
@@ -58,7 +58,7 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
         // NOTE: Safari 13 blocks attempts to focus elements inside a third-party iframe before the user interacts with it
         // https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes
         // We can work around this restriction by focusing the <iframe> element beforehand
-        if (isWindowInIframe(window))
+        if (isIframeWindow(window))
             await sendRequestToFrame({ cmd: GET_IFRAME_REQUEST_CMD }, GET_IFRAME_RESPONSE_CMD, window.parent);
 
         const activeElement               = domUtils.getActiveElement();

--- a/src/client/automation/utils/utils.js
+++ b/src/client/automation/utils/utils.js
@@ -58,7 +58,7 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
         // NOTE: Safari 13 blocks attempts to focus elements inside a third-party iframe before the user interacts with it
         // https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes
         // We can work around this restriction by focusing the <iframe> element beforehand
-        if (isWindowInIframe(window)())
+        if (isWindowInIframe(window))
             await sendRequestToFrame({ cmd: GET_IFRAME_REQUEST_CMD }, GET_IFRAME_RESPONSE_CMD, window.parent);
 
         const activeElement               = domUtils.getActiveElement();

--- a/src/client/automation/utils/utils.js
+++ b/src/client/automation/utils/utils.js
@@ -1,6 +1,7 @@
 import hammerhead from '../deps/hammerhead';
 import testCafeCore from '../deps/testcafe-core';
 import sendRequestToFrame from '../../core/utils/send-request-to-frame';
+import isWindowInIframe from '../../../utils/is-window-in-iframe';
 
 const Promise          = hammerhead.Promise;
 const nativeMethods    = hammerhead.nativeMethods;
@@ -57,7 +58,7 @@ export function focusAndSetSelection (element, simulateFocus, caretPos) {
         // NOTE: Safari 13 blocks attempts to focus elements inside a third-party iframe before the user interacts with it
         // https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes
         // We can work around this restriction by focusing the <iframe> element beforehand
-        if (window.top !== window)
+        if (isWindowInIframe(window)())
             await sendRequestToFrame({ cmd: GET_IFRAME_REQUEST_CMD }, GET_IFRAME_RESPONSE_CMD, window.parent);
 
         const activeElement               = domUtils.getActiveElement();

--- a/src/client/automation/utils/utils.js
+++ b/src/client/automation/utils/utils.js
@@ -1,5 +1,6 @@
 import hammerhead from '../deps/hammerhead';
 import testCafeCore from '../deps/testcafe-core';
+import sendRequestToFrame from '../../core/utils/send-request-to-frame';
 
 const Promise          = hammerhead.Promise;
 const nativeMethods    = hammerhead.nativeMethods;
@@ -11,6 +12,20 @@ const textSelection   = testCafeCore.textSelection;
 const domUtils        = testCafeCore.domUtils;
 const styleUtils      = testCafeCore.styleUtils;
 
+const messageSandbox = hammerhead.eventSandbox.message;
+
+const GET_IFRAME_REQUEST_CMD  = 'automation|iframe|request';
+const GET_IFRAME_RESPONSE_CMD = 'automation|iframe|response';
+
+messageSandbox.on(messageSandbox.SERVICE_MSG_RECEIVED_EVENT, e => {
+    if (e.message.cmd === GET_IFRAME_REQUEST_CMD) {
+        const iframeElement = domUtils.findIframeByWindow(e.source);
+
+        focusBlurSandbox.focus(iframeElement, () => {
+            messageSandbox.sendServiceMsg({ cmd: GET_IFRAME_RESPONSE_CMD }, e.source);
+        }, false);
+    }
+});
 
 function setCaretPosition (element, caretPos) {
     const isTextEditable    = domUtils.isTextEditableElement(element);
@@ -38,7 +53,13 @@ function setCaretPosition (element, caretPos) {
 }
 
 export function focusAndSetSelection (element, simulateFocus, caretPos) {
-    return new Promise(resolve => {
+    return new Promise(async resolve => {
+        // NOTE: Safari 13 blocks attempts to focus elements inside a third-party iframe before the user interacts with it
+        // https://developer.apple.com/documentation/safari-release-notes/safari-13-release-notes
+        // We can work around this restriction by focusing the <iframe> element beforehand
+        if (window.top !== window)
+            await sendRequestToFrame({ cmd: GET_IFRAME_REQUEST_CMD }, GET_IFRAME_RESPONSE_CMD, window.parent);
+
         const activeElement               = domUtils.getActiveElement();
         const isTextEditable              = domUtils.isTextEditableElement(element);
         const labelWithForAttr            = domUtils.closest(element, 'label[for]');

--- a/src/client/driver/command-executors/browser-manipulation/index.js
+++ b/src/client/driver/command-executors/browser-manipulation/index.js
@@ -163,7 +163,7 @@ class ManipulationExecutor {
     }
 
     _requestManipulation () {
-        if (!isWindowInIframe(window)())
+        if (!isWindowInIframe(window))
             return transport.queuedAsyncServiceMsg(this._createManipulationReadyMessage());
 
         const cropDimensions = this._getAbsoluteCropValues();
@@ -202,7 +202,7 @@ class ManipulationExecutor {
                 return this._runScrollBeforeScreenshot();
             })
             .then(() => {
-                if (!isWindowInIframe(window)())
+                if (!isWindowInIframe(window))
                     return this._hideUI();
 
                 return Promise.resolve();
@@ -216,7 +216,7 @@ class ManipulationExecutor {
 
                 manipulationResult = result;
 
-                if (!isWindowInIframe(window)())
+                if (!isWindowInIframe(window))
                     this._showUI();
 
                 return delay(POSSIBLE_RESIZE_ERROR_DELAY);

--- a/src/client/driver/command-executors/browser-manipulation/index.js
+++ b/src/client/driver/command-executors/browser-manipulation/index.js
@@ -26,6 +26,7 @@ import runWithBarriers from '../../utils/run-with-barriers';
 import MESSAGE from '../../../../test-run/client-messages';
 import COMMAND_TYPE from '../../../../test-run/commands/type';
 import { ScrollOptions, ElementScreenshotOptions } from '../../../../test-run/commands/options';
+import isWindowInIframe from '../../../../utils/is-window-in-iframe';
 
 
 const messageSandbox = eventSandbox.message;
@@ -162,7 +163,7 @@ class ManipulationExecutor {
     }
 
     _requestManipulation () {
-        if (window.top === window)
+        if (!isWindowInIframe(window)())
             return transport.queuedAsyncServiceMsg(this._createManipulationReadyMessage());
 
         const cropDimensions = this._getAbsoluteCropValues();
@@ -201,7 +202,7 @@ class ManipulationExecutor {
                 return this._runScrollBeforeScreenshot();
             })
             .then(() => {
-                if (window.top === window)
+                if (!isWindowInIframe(window)())
                     return this._hideUI();
 
                 return Promise.resolve();
@@ -215,7 +216,7 @@ class ManipulationExecutor {
 
                 manipulationResult = result;
 
-                if (window.top === window)
+                if (!isWindowInIframe(window)())
                     this._showUI();
 
                 return delay(POSSIBLE_RESIZE_ERROR_DELAY);

--- a/src/client/driver/command-executors/browser-manipulation/index.js
+++ b/src/client/driver/command-executors/browser-manipulation/index.js
@@ -26,7 +26,7 @@ import runWithBarriers from '../../utils/run-with-barriers';
 import MESSAGE from '../../../../test-run/client-messages';
 import COMMAND_TYPE from '../../../../test-run/commands/type';
 import { ScrollOptions, ElementScreenshotOptions } from '../../../../test-run/commands/options';
-import isWindowInIframe from '../../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../../utils/is-window-in-iframe';
 
 
 const messageSandbox = eventSandbox.message;
@@ -163,7 +163,7 @@ class ManipulationExecutor {
     }
 
     _requestManipulation () {
-        if (!isWindowInIframe(window))
+        if (!isIframeWindow(window))
             return transport.queuedAsyncServiceMsg(this._createManipulationReadyMessage());
 
         const cropDimensions = this._getAbsoluteCropValues();
@@ -202,7 +202,7 @@ class ManipulationExecutor {
                 return this._runScrollBeforeScreenshot();
             })
             .then(() => {
-                if (!isWindowInIframe(window))
+                if (!isIframeWindow(window))
                     return this._hideUI();
 
                 return Promise.resolve();
@@ -216,7 +216,7 @@ class ManipulationExecutor {
 
                 manipulationResult = result;
 
-                if (!isWindowInIframe(window))
+                if (!isIframeWindow(window))
                     this._showUI();
 
                 return delay(POSSIBLE_RESIZE_ERROR_DELAY);

--- a/src/client/ui/status-bar/index.js
+++ b/src/client/ui/status-bar/index.js
@@ -224,7 +224,7 @@ export default class StatusBar extends serviceUtils.EventEmitter {
     }
 
     _createBeforeReady () {
-        if (this.state.created || isWindowInIframe(window)())
+        if (this.state.created || isWindowInIframe(window))
             return;
 
         if (document.body)

--- a/src/client/ui/status-bar/index.js
+++ b/src/client/ui/status-bar/index.js
@@ -3,6 +3,7 @@ import testCafeCore from './../deps/testcafe-core';
 import ProgressBar from './progress-bar';
 import uiRoot from '../ui-root';
 import MESSAGES from './messages';
+import isWindowInIframe from '../../../utils/is-window-in-iframe';
 
 
 const Promise          = hammerhead.Promise;
@@ -223,7 +224,7 @@ export default class StatusBar extends serviceUtils.EventEmitter {
     }
 
     _createBeforeReady () {
-        if (this.state.created || window !== window.top)
+        if (this.state.created || isWindowInIframe(window)())
             return;
 
         if (document.body)

--- a/src/client/ui/status-bar/index.js
+++ b/src/client/ui/status-bar/index.js
@@ -3,7 +3,7 @@ import testCafeCore from './../deps/testcafe-core';
 import ProgressBar from './progress-bar';
 import uiRoot from '../ui-root';
 import MESSAGES from './messages';
-import isWindowInIframe from '../../../utils/is-window-in-iframe';
+import isIframeWindow from '../../../utils/is-window-in-iframe';
 
 
 const Promise          = hammerhead.Promise;
@@ -224,7 +224,7 @@ export default class StatusBar extends serviceUtils.EventEmitter {
     }
 
     _createBeforeReady () {
-        if (this.state.created || isWindowInIframe(window))
+        if (this.state.created || isIframeWindow(window))
             return;
 
         if (document.body)

--- a/src/utils/is-window-in-iframe.ts
+++ b/src/utils/is-window-in-iframe.ts
@@ -1,3 +1,3 @@
-export default function isWindowInIframe (window: Window): boolean {
+export default function isIframeWindow (window: Window): boolean {
     return window.top !== window;
 }

--- a/src/utils/is-window-in-iframe.ts
+++ b/src/utils/is-window-in-iframe.ts
@@ -1,0 +1,3 @@
+export default function isWindowInIframe (window: Window): boolean {
+    return window.top !== window;
+}

--- a/test/client/fixtures/automation/type-test.js
+++ b/test/client/fixtures/automation/type-test.js
@@ -423,7 +423,7 @@ $(document).ready(function () {
     asyncTest('GH-4068 - Type to element wrapped in label', function () {
         const input1 = document.createElement('input');
         const input2 = document.createElement('input');
-        const label  =  document.createElement('label');
+        const label  = document.createElement('label');
 
         input1.className = TEST_ELEMENT_CLASS;
         input2.className = TEST_ELEMENT_CLASS;

--- a/test/functional/fixtures/regression/gh-4793/pages/iframe.html
+++ b/test/functional/fixtures/regression/gh-4793/pages/iframe.html
@@ -1,0 +1,8 @@
+<!DOCTYPE html>
+<html>
+<head>
+</head>
+<body>
+    <input id="input"/>
+</body>
+</html>

--- a/test/functional/fixtures/regression/gh-4793/pages/index.html
+++ b/test/functional/fixtures/regression/gh-4793/pages/index.html
@@ -1,0 +1,9 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>GH-4793</title>
+</head>
+<body>
+<iframe id="cross-domain-iframe" src="http://localhost:3001/fixtures/regression/gh-4793/pages/iframe.html"></iframe>
+</body>
+</html>

--- a/test/functional/fixtures/regression/gh-4793/test.js
+++ b/test/functional/fixtures/regression/gh-4793/test.js
@@ -1,0 +1,5 @@
+describe('[Regression](GH-4793)', function () {
+    it('Elements inside cross-domain iframes should be focusable', function () {
+        return runTests('testcafe-fixtures/index-test.js', 'Type text into an input inside a cross-domain iframe');
+    });
+});

--- a/test/functional/fixtures/regression/gh-4793/testcafe-fixtures/index-test.js
+++ b/test/functional/fixtures/regression/gh-4793/testcafe-fixtures/index-test.js
@@ -4,7 +4,7 @@ fixture `gh-4793`
     .page `http://localhost:3000/fixtures/regression/gh-4793/pages/index.html`;
 
 test('Type text into an input inside a cross-domain iframe', async t => {
-    const iframe = Selector('#cross-domain-iframe');
+    const iframe = Selector('#cross-domain-iframe', { timeout: 10000 });
     const input  = Selector('#input');
 
     await t

--- a/test/functional/fixtures/regression/gh-4793/testcafe-fixtures/index-test.js
+++ b/test/functional/fixtures/regression/gh-4793/testcafe-fixtures/index-test.js
@@ -1,0 +1,14 @@
+import { Selector } from 'testcafe';
+
+fixture `gh-4793`
+    .page `http://localhost:3000/fixtures/regression/gh-4793/pages/index.html`;
+
+test('Type text into an input inside a cross-domain iframe', async t => {
+    const iframe = Selector('#cross-domain-iframe');
+    const input  = Selector('#input');
+
+    await t
+        .switchToIframe(iframe)
+        .typeText(input, '1234')
+        .expect(input.value).eql('1234');
+});

